### ISSUE TITLE
a11y: hit test order when two semantic nodes are overlapping

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#11668]
Setup: A floating app bar with a long list. The list is half-way scrolled down. The AppBar is in its smallest state. TalkBack is turned on. 

It is now impossible to select anything by tapping on it in the app bar. The a11y focus always goes to  the list item hidden behind the app bar.

Related: https://github.com/flutter/flutter/issues/11664

**Comment by goderbauer** [comment]
This problem only exists on Android. Works fine on iOS.

**Comment by goderbauer** [comment]
Actually, in certain situations, iOS also seems to have the same problem.

**Comment by goderbauer** [comment]
On Android the problem is as follows:

In the SemanticsNode hit test implementation we assume that children are present in paint order (and the hit test order is the reverse of that). However, `visitChildrenForSemantics` on the `RenderViewport` does not return its children in paint order even though it should. That's why in the SemanticsNode hit test implementation the wrong node wins and gets the focus.

Easy fix for this is to fix `visitChildrenForSemantics` on the viewport to return children in paint order. However, that breaks the traversal order, which is different.

It also makes the semantics tree awkward to read. Imagine we have an app bar with a title, a list with 3 items and another sliver as footer at the end. The tree would now include the semantic nodes in a hard to read order:

* Footer
* List Item 1
* List Item 2
* List Item 3
* App Bar Title

It cannot be read from top to bottom anymore. For this example, you'll start at the bottom, then read the middle and then the top. It gets even worse for Viewports with widgets on both side of a center.

**Comment by Hixie** [comment]
I wouldn't worry too much about the semantics tree being easy to read. If that's really a problem we can always make the semantics tree printer print in traversal order.
